### PR TITLE
Backend

### PR DESCRIPTION
**Pull request by GustavHandmark**
Merged algorithm branch into backend branch locally 10 oct. 21:15.

Alot of new changes introduced. Trip class and helper classes (tripResources & tripDataAccess) should work with frontend. No API documentation available at this point, still work in progress (but shouldn't be hard to figure out from the testcases available). Java doc still in progress. We figured frontend would want to work against the backend as soon as possible to try real-case scenarios.

### CODE REVIEW

File: src/main/resources/se/lth/base/server/database/schema.sql [hunk 1]
```diff
@@ -25,42 +26,60 @@ CREATE TABLE session(session_uuid UUID DEFAULT RANDOM_UUID(),
...
 
-    PRIMARY KEY(foo_id),
-    -- This defines foo_id as the unique identifier of the table. It adds NOT NULL to the column and
+    PRIMARY KEY(tripId),
+    -- This defines trip_id as the unique identifier of the table. It adds NOT NULL to the column and
     -- enforces that the values rows all have a unique identifier.
+    --
 
-    FOREIGN KEY(user_id) REFERENCES user(user_id) ON DELETE CASCADE
+    -- Maybe we should add a foreign key to link the tables?
```
Comment (jadlers) [inline src/main/resources/se/lth/base/server/database/schema.sql]: `TODO: `?

File: src/test/java/se/lth/base/server/rest/TripResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    assertTrue(co2response.compareTo(expected) == 0);
+    Response complete = target("trip").path("1/complete").request().get();
+    Response rpAfterComplete = target("trip").path("1/saved").request().get();
+    Trip tripData = target("trip").path("1").request().get(Trip.class);
+    Float co2aftercomplete = new Float(rpAfterComplete.readEntity(Float.class));
+    assertTrue(co2aftercomplete.compareTo(new Float(tripData.co2SavedCalc())) == 0);
+
+    // System.out.println(new String(complete.readEntity(String.class)));
+    // System.out.println(new String(rp.readEntity(String.class)));
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/rest/TripResourceTest.java]: Remove unnecessary `System.out.println`

File: src/test/java/se/lth/base/server/rest/TripResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    /*
+    JsonObject jsobj = (JsonObject) new Gson().toJsonTree(newTrip);
+    Trip testWithJson = new Gson().fromJson(jsobj,Trip.class);
+    assertEquals(testWithJson.toString(),newTripCreate.toString());
+    //System.out.println("Gson Test:\n" + testWithJson);
+    Trip testRequestWithJson = target("trip").request().post(Entity.json(jsobj),Trip.class);
+    //System.out.println("last test\n"+testRequestWithJson);
+    assertEquals(testRequestWithJson.toString(),newTripCreate.toString());
+    // System.out.println(responseTrip);
+    */
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/rest/TripResourceTest.java]: Lots of commented out code, remove?

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 4]
```diff
@@ -20,8 +20,9 @@
 
   // 1,1,(4,5,6),5,3,'2018-10-08T08:00','2018-10-08T08:45',(55.7029296,13.1929449,55.6052931,13.0001566),'Lund,
   // Sverige','Malmö, Sverige',22.5,FALSE)
-  @Test
-  public void addTrip() {
+
+  @Before
+  public void initiaialize() {
```
Comment (FranzGithub) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: initialize*

File: src/main/java/se/lth/base/server/rest/UserResource.java [hunk 5]
```diff
@@ -135,6 +135,17 @@ public User putUser(@PathParam("id") int userId, Credentials credentials) {
     return userDao.updateUser(userId, credentials);
   }
 
+  @Path("{id}/reset_password")
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON + ";charset=utf-8")
+  @PermitAll
+  public User resetPassword(@PathParam("id") int userId, Credentials credentials) {
```
Comment (jadlers) [inline src/main/java/se/lth/base/server/rest/UserResource.java]: This opens up for anyone to change a users password using only an `id`. Not that the security is a high priority but what happened to the receiving an email approach?

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 6]
```diff
@@ -43,14 +44,55 @@ public void addTrip() {
             "Lund, Sverige",
             22.5,
             false);
-    assertEquals(TEST.getId(), data.getDriver());
-    assertTrue(tripDao.getTripFromId(3) != null);
+  }
+
+  @Test
+  public void addTrip() {
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: `addTrip` does not test adding a trip

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 7]
```diff
@@ -43,14 +44,55 @@ public void addTrip() {
...
+    //            passengers,
+    //            5,
+    //            3,
+    //            "2018-10-8T08:00",
+    //            "2018-10-8T08:45",
+    //            coordinates,
+    //            "Malmö, Sverige",
+    //            "Lund, Sverige",
+    //            22.5,
+    //            false);
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: Remove all commented out code

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 8]
```diff
@@ -43,14 +44,55 @@ public void addTrip() {
...
+    assertEquals(TEST.getId(), tripDao.getTripFromId(1).getDriver());
+    assertTrue(tripDao.getTripFromId(1) != null);
     // tripDao.getAllTrips().forEach((trip) -> System.out.println(trip));
   }
 
   // two trips added by default
   @Test
+  // @Ignore // TODO: Fix this test and remove @Ignore DONE:
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: Remove `TODO` comment since it's done

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 9]
```diff
@@ -43,14 +44,55 @@ public void addTrip() {
...
+        coordinates,
+        "Malmö, Sverige",
+        "Lund, Sverige",
+        24,
+        false);
+    assertEquals(2, tripDao.getAllTrips().size());
+
     /*
      * tripDao.addTrip( TEST.getId(), "2018-10-8T08:00", "2018-10-8T08:45",
      * "Malmö, Sverige", "Lund, Sverige", 22.5, false); assertEquals(3,
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: Remove this commented out code

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 10]
```diff
@@ -62,38 +104,63 @@ public void getAllTrips() {
 
   }
 
+  /** TODO: The test shall not assume anything other than what's in the @Before method */
   @Test
+  // @Ignore // TODO: Remove
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: Remove `TODO` comment

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 11]
```diff
@@ -62,38 +104,63 @@ public void getAllTrips() {
...
-    assertEquals(2, tripDao.getUsersTrips(1).size());
+    assertEquals(1, tripDao.getUsersTrips(TEST.getId()).size());
+    // add another trip
+    passengers.add(6);
+    passengers.add(7);
+    passengers.add(8);
+    coordinates.add(55.7029296f);
+    coordinates.add(13.1929449f);
+    coordinates.add(55.6052931f);
+    coordinates.add(13.0001566f);
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: Why add more values to `passengers` and `coordinates`? They are initialised in the `@Before` method.

File: src/test/java/se/lth/base/server/data/TripDataAccessTest.java [hunk 12]
```diff
@@ -62,38 +104,63 @@ public void getAllTrips() {
...
-    assertEquals(2, tripDao.getAllTrips().size());
+    assertEquals(1, tripDao.getAllTrips().size());
+    // add another trip
     passengers.add(3);
     passengers.add(4);
     passengers.add(5);
     coordinates.add(55.7029296f);
     coordinates.add(13.1929449f);
     coordinates.add(55.6052931f);
     coordinates.add(13.0001566f);
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripDataAccessTest.java]: Why add more values to `passengers` and `coordinates`? They are initialised in the `@Before` method.

File: src/test/java/se/lth/base/server/data/TripTest.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  @Test
+  public void getco2SavedTest() {
+    assertEquals(0, data.getco2Saved(), 0.0001);
+    tripDao.setTripCompleted(data.getTripId());
+    Trip t = tripDao.getTripFromId(data.getTripId());
+    assertFalse(0 == t.getco2Saved());
+  }
+
+  @Test
+  public void toStringTest() {
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/TripTest.java]: Not an actual test

File: src/test/java/se/lth/base/server/data/UserDataAccessTest.java [hunk 14]
```diff
@@ -160,4 +160,14 @@ public void updateWithoutPassword() {
     System.out.println(session1);
     System.out.println(session2);
   }
+
+  @Test
+  public void updatePassword() {
+    User og = userDao.addUser(new Credentials("Pelle", "ogPassword", Role.USER));
+    userDao.updatePassword(og.getId(), new Credentials("ngPassword"));
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/data/UserDataAccessTest.java]: Don't create a new contructor for `Credentials` for this one test

Comment (jadlers) [inline src/test/java/se/lth/base/server/data/UserDataAccessTest.java]: Don't create a new constructor for `Credentials` for this one test

File: src/test/java/se/lth/base/server/rest/TripResourceTest.java [hunk 15]
```diff
@@ -90,67 +121,22 @@ public void getCo2Saved() {
...
-    // System.out.println(new String(rp.readEntity(String.class)));
-
   }
 
   @Test
   public void getAllTrips() {
-    login(ADMIN_CREDENTIALS);
+    target("trip").request().post(Entity.json(expectedTrip), Trip.class);
+    target("trip").request().post(Entity.json(newTripCreate), Trip.class);
     Response rp = target("trip").path("all").request().get();
```
Comment (jadlers) [inline src/test/java/se/lth/base/server/rest/TripResourceTest.java]: This `Response rp` is never used, delete

File: src/main/java/se/lth/base/server/data/Credentials.java [hunk 16]
```diff
@@ -37,6 +37,10 @@ public Credentials(String username, String password, Role role) {
     this(username, null, password, role);
   }
 
+  public Credentials(String password) {
```
[the remark this answers is not in the document]
Comment (jadlers) [reply src/main/java/se/lth/base/server/data/Credentials.java]: I found where it's used and have left a comment there as well.

TL;DR: Remove this constructor

File: src/main/java/se/lth/base/server/rest/TripResource.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  @GET
+  @RolesAllowed(Role.Names.USER)
+  @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
+  public float getCo2Saved(@PathParam("id") int tripId) {
+    return tripDao.calculateCo2Saved(tripId);
+  }
+
+  // Should be done in a scheduler -- fix if we have time.
+  @Path("{id}/complete")
+  @GET
```
Comment (jadlers) [inline src/main/java/se/lth/base/server/rest/TripResource.java]: Should this not be a `@POST` request?
